quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+            self.parsed_artifact_name = package_file_name
+            self.parsed_pkg = "%s/%s/%s" % (package_name, package_version, package_file_name)
+            self.parsed_endpoint_url = "/projects/%s/packages/generic/%s" % (
+                urllib.parse.quote_plus(self.parsed_project_name),
+                self.parsed_pkg,
+            )
+        else:
+            raise QuarkError("Unsupported URL: " + url)
+
+    def _resolve_job_id(self, project: str, ref: str, job_name: str) -> str:
+        # NOTE: This method requires an API token. A CI job token isn't sufficient. As such, this
+        # should be called only when freezing (usually done by a developer with a private token).
+
+        # Latest pipeline for given ref
+        req = urllib.request.Request(
+            url="%s/api/v4/projects/%s/pipelines/latest?ref=%s" % (
+                self.gitlab_url,
+                urllib.parse.quote_plus(project),
+                urllib.parse.quote_plus(ref),
+            ),
+            headers={
+                self.gitlab_token_header: self.gitlab_token,
+            },
+        )
+
+        with urllib.request.urlopen(req) as response:
+            pipeline = json.loads(response.read().decode("utf-8"))
+
+        # Jobs in latest pipeline
+        req = urllib.request.Request(
+            url="%s/api/v4/projects/%s/pipelines/%s/jobs" % (
+                self.gitlab_url,
+                urllib.parse.quote_plus(project),
+                pipeline["id"],
+            ),
+            headers={
+                self.gitlab_token_header: self.gitlab_token,
+            },
+        )
+
+        with urllib.request.urlopen(req) as response:
+            jobs = json.loads(response.read().decode("utf-8"))
+
+        for job in jobs:
+            if job["name"] == job_name:
+                return str(job["id"])
+
+        raise QuarkError("Could not find job %s in pipeline %s" % (job_name, pipeline["id"]))
+
+    def _download(self, tempdir: str) -> str:
+        assert self.parsed_artifact_name  # Make Pyright happy
+        assert self.parsed_endpoint_url  # Make Pyright happy
+
+        print_msg("downloading " + self.parsed_artifact_name, self._comment())
+
+        dl_dir = join(tempdir, "dl")
+        archive_path = join(dl_dir, self.parsed_artifact_name)
+
+        os.mkdir(dl_dir)
+
+        req = urllib.request.Request(
+            url="%s/api/v4/%s" % (self.gitlab_url, self.parsed_endpoint_url.lstrip("/")),
+            headers={
+                self.gitlab_token_header: self.gitlab_token,
+            },
+        )
+
+        with urllib.request.urlopen(req) as response:
+            with open(archive_path, "wb") as fileobj:
+                shutil.copyfileobj(response, fileobj)
+
+        # Verify or update the SHA1
+        print_msg("verifying " + self.parsed_artifact_name, self._comment())
+        with open(archive_path, "rb") as fileobj:
+            sha1 = hashlib.sha1(fileobj.read()).hexdigest()
```

Comment:
We are downloading stuff that is potentially big, you don't want to read it all in memory to compute the hash. You probably want something like:

```python
            sha1 = hashlib.file_digest(fileobj, "sha1").hexdigest()
```

Reply:
(as in the following commit you switch to reading by chunks explicitly, you may possibly even update the hash in there to avoid re-reading the file)